src/redis-cli.c

```diff
@@ -1892,8 +2027,23 @@ static int cliSendCommand(int argc, char **argv, long repeat) {
                 config.dbnum = 0;
                 config.conn_info.input_dbnum = 0;
                 config.resp3 = 0;
+                if (config.pubsub_mode && config.push_output) {
+                    redisSetPushCallback(context, cliPushHandler);
+                }
+                config.pubsub_mode = 0;
                 cliRefreshPrompt();
+            } else if (!strcasecmp(command,"hello")) {
+                if (config.last_cmd_type == REDIS_REPLY_MAP)
+                    config.resp3 = 2;
```

Comment:
Good points. I assumed users only type HELLO if they want to switch to RESP3, but you're right it can be called even if RESP3 is already active.

The values 1 (explicit) and 2 (implicit) seemed vague to me. Looking at how the values are used, I realize 2 means it can be overridden by -2 (for example `--json -2`  means use RESP2) and also it is not an error if the initial `HELLO 3` fails if resp3 == 2, which I suppose it to be able support older Redis versions where RESP3 was not implemented (and HELLO is an unknown command).

Here after HELLO 3 has returned a map, we know that we have RESP3 so setting it to 1 seems right. I think we can do it regardless of the previous value (0, 1 or 2) but I can do it only `if (config.resp3 == 0)` if you think it's better.

Reply:
that code also run when we get the response of an implicit HELLO 3 we sent (i.e. when user just passed `--json`), right?

looking closer, i see this variable is actually tied to command line arguments, and controls which HELLO we'll send when connecting / reconnecting.
so i wonder if at all we should modify it post startup (also for RESET, which already modifies it).
maybe instead we should have another variable that's indicating the current connection's RESP mode (separate from the user's intention when starting redis-cli, which controls the (re)connection?

Reply:
> that code also run when we get the response of an implicit HELLO 3 we sent (i.e. when user just passed `--json`), right?

No. This code only runs on user input HELLO.

> looking closer, i see this variable is actually tied to command line arguments, and controls which HELLO we'll send when connecting / reconnecting.
> so i wonder if at all we should modify it post startup (also for RESET, which already modifies it).
> maybe instead we should have another variable that's indicating the current connection's RESP mode (separate from the user's intention when starting redis-cli, which controls the (re)connection?

Is it really that important? The only difference is that you get RESP3 again after reconnect, rather than being reset to what you had when connecting, which is probably what the user wants anyway if they've explicitly issued a HELLO command.

Reply:
well, ok, i suppose this is the path of least surprise.

Reply:
One scenario I can think of: A new redis-cli connects to an old redis-server without RESP3 support. The user specifies e.g. `--json` so config.resp3 == 2. This doesn't cause cliSwitchProto to fail, so the user ends up connected with RESP2. The problem then is the check for pubsub messages.

```C
int isPubsubPush(redisReply *r) {
    if (r == NULL ||
        r->type != (config.resp3 ? REDIS_REPLY_PUSH : REDIS_REPLY_ARRAY) ||
...
```

To support this scenario, I guess we need to know if we actually have RESP3 right now or not.

Reply:
Now I added a `config.current_resp3` that keeps track of the current state, as you suggested some hours ago.